File: app/src/main/java/com/github/dedis/student20_pop/model/network/level/high/meeting/StateMeeting.java
package com.github.dedis.student20_pop.model.network.level.high.meeting;

import com.github.dedis.student20_pop.model.network.level.high.Action;
import com.github.dedis.student20_pop.model.network.level.high.Message;
import com.github.dedis.student20_pop.model.network.level.high.Objects;

public class StateMeeting extends Message {

    private final String id; // Hash(lao_id + creation + name)
    private final String name;
    private final long creation;
    private final long last_modified;
    private final String location;
    private final long start;
    private final long end;

    public StateMeeting(String id, String name, long creation, long last_modified, String location, long start, long end) {
        this.id = id;
        this.name = name;
        this.creation = creation;
        this.last_modified = last_modified;
        this.location = location;
        this.start = start;
        this.end = end;
    }
    //private final Extra extra;

    @Override
    public String getObject() {
        return Objects.MEETING.getObject();
    }

    @Override
    public String getAction() {
        return Action.STATE.getAction();
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public long getCreation() {
        return creation;
    }

    public long getLast_modified() {
        return last_modified;
    }

    public String getLocation() {
        return location;
    }

    public long getStart() {
        return start;
    }

    public long getEnd() {
        return end;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        StateMeeting that = (StateMeeting) o;
        return getCreation() == that.getCreation() &&
                getLast_modified() == that.getLast_modified() &&
                getStart() == that.getStart() &&
                getEnd() == that.getEnd() &&
                java.util.Objects.equals(getId(), that.getId()) &&
                java.util.Objects.equals(getName(), that.getName()) &&
                java.util.Objects.equals(getLocation(), that.getLocation());
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(getId(), getName(), getCreation(), getLast_modified(), getLocation(), getStart(), getEnd());
    }
}

File: fe1-web/Model/Network/Method/Message/data/meeting/StateMeeting.ts
import {Base64Data, Hash, Lao, PublicKey, Signature, Timestamp} from "Model/Objects";
import { ActionType, MessageData, ObjectType } from "../messageData";
import { WitnessSignature } from "Model/Objects/WitnessSignature";
import { ProtocolError } from "../../../../ProtocolError";
import { checkModificationId, checkModificationSignatures, checkTimestampStaleness } from "../checker";
import { OpenedLaoStore } from 'Store';
import { eventTags } from "../../../../../../websockets/WebsocketUtils";

export class StateMeeting implements MessageData {

  public readonly object: ObjectType = ObjectType.MEETING;
  public readonly action: ActionType = ActionType.STATE;

  public readonly id: Hash;
  public readonly name: string;
  public readonly creation: Timestamp;
  public readonly last_modified: Timestamp;
  public readonly location?: string;
  public readonly start: Timestamp;
  public readonly end?: Timestamp;
  public readonly extra?: {};
  public readonly modification_id: Hash;
  public readonly modification_signatures: WitnessSignature[];

  constructor(msg: Partial<StateMeeting>) {

    if (!msg.name)
      throw new ProtocolError('Undefined \'name\' parameter encountered during \'StateMeeting\'');
    this.name = msg.name;

    if (!msg.creation)
      throw new ProtocolError('Undefined \'creation\' parameter encountered during \'StateMeeting\'');
    checkTimestampStaleness(msg.creation);
    this.creation = new Timestamp(msg.creation.toString());

    if (!msg.last_modified)
      throw new ProtocolError('Undefined \'last_modified\' parameter encountered during \'StateMeeting\'');
    if (msg.last_modified < msg.creation)
      throw new ProtocolError('Invalid timestamp encountered: \'last_modified\' parameter smaller than \'creation\'');
    this.last_modified = new Timestamp(msg.last_modified.toString());

    if (msg.location) this.location = msg.location;

    if (!msg.start)
      throw new ProtocolError('Undefined \'start\' parameter encountered during \'StateMeeting\'');
    checkTimestampStaleness(msg.start);
    this.start = new Timestamp(msg.start.toString());

    if (msg.end) {
      if (msg.end < msg.creation)
        throw new ProtocolError('Invalid timestamp encountered: \'end\' parameter smaller than \'creation\'');
      if (msg.end < msg.start)
        throw new ProtocolError('Invalid timestamp encountered: \'end\' parameter smaller than \'start\'');
      this.end = new Timestamp(msg.end.toString());
    }

    if (msg.extra) this.extra = JSON.parse(JSON.stringify(msg.extra)); // clone JS object extra

    if (!msg.modification_id)
      throw new ProtocolError('Undefined \'modification_id\' parameter encountered during \'StateMeeting\'');
    checkModificationId(msg.modification_id);
    this.modification_id = new Hash(msg.modification_id.toString());

    if (!msg.modification_signatures)
      throw new ProtocolError('Undefined \'modification_signatures\' parameter encountered during \'StateMeeting\'');
    checkModificationSignatures(msg.modification_signatures);
    this.modification_signatures = msg.modification_signatures.map((ws) =>
      new WitnessSignature({
        witness: new PublicKey(ws.witness.toString()),
        signature: new Signature(ws.signature.toString()),
      })
    );

    if (!msg.id) throw new ProtocolError('Undefined \'id\' parameter encountered during \'StateMeeting\'');
    const lao: Lao = OpenedLaoStore.get();
    /* // FIXME get event from storage
    const expectedHash = Hash.fromStringArray(eventTags.MEETING, lao.id.toString(), lao.creation.toString(), MEETING_NAME);
    if (!expectedHash.equals(msg.id))
      throw new ProtocolError('Invalid \'id\' parameter encountered during \'StateMeeting\': unexpected id value'); */
    this.id = new Hash(msg.id.toString());
  }

  public static fromJson(obj: any): StateMeeting {

    // FIXME add JsonSchema validation to all "fromJson"
    let correctness = true;

    return correctness
    ? new StateMeeting(obj)
    : (() => { throw new ProtocolError("add JsonSchema error message"); })();
  }
}
